### ISSUE TITLE
Prebuilt Provider Packages

### ISSUE THREAD

**Issue by skorfmann** [hashicorp/terraform-cdk#98]
When thinking about distributing L2 or L3 constructs, we'll need provider packages. Without, it's quite hard to define the dependency on used providers (e.g. Null Provider). 

![CleanShot 2020-06-07 at 15 27 38](https://user-images.githubusercontent.com/136789/84129593-75d9e880-aa42-11ea-974f-ccfc4fef18ae.png)

## Current State 

In order to do this at the moment by the example of `aws-ecr-asset`:

- Create / Publish `aws-ecr-asset` package
- Add a note in the Readme / post install to add the `null` provider to `cdktf.json` 
- When there's a breaking change in one of the upstream packages, it'll break and the user will have dig through weird error messages

## Desired State

We have prebuilt provider packages, with potentially a few predefined L2 / L3 constructs for each of these providers, which allows leaving the dependency management to proper tooling (npm et al). This would also get us closer to a workflow where there's essentially no need for a local Terrafor CLI (if the Terraform Cloud workflow is chosen).

The version of distributed packages shouldn't be tied to the original provider, but always track latest. The version would be mandated by changes to `cdktf` the `<provider>-constructs` and possibly breaking changes by the actual underlying Terraform Provider / Engine.

- The provider package gets auto-generated via `cdktf get`
- It tracks `latest` of the provider
- Users which require a specific provider version for some reason, will have to fallback to generate it manually by `cdktf get`

### Proposed Repo Structure 

To allow individiual versioning of providers, I think it makes sense to use something like the following repo structure

#### hashicorp/terraform-cdk

- packages/cdktf
- packages/cdktf-cli

#### hashicorp/terraform-cdk-aws

- packages/@cdktf/aws # metapackage to pull in cdktf / aws provider / aws constructs
- packages/@cdktf/aws-provider
- packages/@cdktf/aws-constructs

#### hashicorp/terraform-cdk-azure

- packages/@cdktf/azure # metapackage to pull in cdktf / azure provider / azure constructs
- packages/@cdktf/azure-provider
- packages/@cdktf/azure-constructs

## Conclusion

Personally, I think that L2 / L3 constructs will be the main selling point for the Terraform CDK. Therefore I'd consider this quite important. 

Thoughts on this one?

**Comment by jsteinich** [comment]
This is something I just started investigating with [Terrastack](https://github.com/jsteinich/terrastack/tree/provider_build). I don't really have much to show, but the idea was to run an action everyday to check for updated provider versions and auto build/publish if any new ones are found.

**Comment by skorfmann** [comment]
> This is something I just started investigating with [Terrastack](https://github.com/jsteinich/terrastack/tree/provider_build). I don't really have much to show, but the idea was to run an action everyday to check for updated provider versions and auto build/publish if any new ones are found.

Yes, that would be totally sufficient for now 👍  I sketched some code already as well. I think at the moment, the primary question is how to do the versioning of these pre-packaged providers. I have some ideas around this, gonna write them up over the weekend.

**Comment by jsteinich** [comment]
I like trying to somehow include the actual provider version, but I agree that matching exactly is likely to be problematic.
Maybe something like **XAA.YBBB.ZC** where **X** = cdktf major version, **AA** = provider major version, **Y** = cdktf minor version, **BBB** = provider minor version, **Z** = cdktf patch version, **C** = provider patch version. Would need to start at 1 for all cdktf versions since semvar calls for no leading zeroes. Would 0 pad provider version numbers as needed to ensure space for future increases.

**Comment by eladb** [comment]
I am wondering if it's sufficient to just somehow have the library relay the "provider requirements" to the app & the CLI instead? Publishing a package for each provider feels like an overkill...

**Comment by skorfmann** [comment]
> I am wondering if it's sufficient to just somehow have the library relay the "provider requirements" to the app & the CLI instead? Publishing a package for each provider feels like an overkill...

@anubhavmishra  and I talked about this, and came to the conclusion that building a custom dependency management wouldn't be worth the effort. I don't think we'd distribute packages for all providers, but a selection of the most used ones ( for now at least).

## Why I think we should do this

### Distribution of custom Constructs

As mentioned above, it's not possible to build custom constructs with dependencies to used providers.

### Speed

I did test building a prepackaged provider for AWS / Azure. This reduced the build time for a `cdktf synth` by ~ 50% (~ 8s / 4s). That's quite significant (related to #151)

### Getting Started Experience

It'd be a much nicer experience if a user wouldn't have to run `cdktf get` to get started.

### Dependency Management / Versioning 

For prepackaged providers we'll have multiple upstream dependencies:

- The respective provider (e.g. AWS or Google) 
- Terraform Engine (indirect via `cdktf`)
- `cdktf` 

I wouldn't bother trying to reflect the individual upstream versions in a prepackaged provider package. I'd rather be opinionated and assume that most users will be fine with `latest`. 

Both, Terraform and the popular providers are quite mature and wouldn't change that often in backwards incompatible ways. For `cdktf` that's certainly different. Anyway, if there would be any breaking upstream changes the packaged provider would just be bumped. 

As already suggested by @jsteinich I think it would make sense to publish new minor or even just patch releases when a new provider got published. This could be run on a schedule, or we somehow subscribe to release notifications.

My assumption is that most users don't care about the specific version of a given provider. It's usually fine to just go with latest. If a user would really care about a certain version, it's still possible to use `cdktf get` to build your own version of that provider as today.

The benefits of this approach:

- No need to build custom logic for things which can be perfectly handled by `npm` et al
- Reduce possible frustration for breaking changes around generated providers
- Users can use the standard tooling to get notified about new updates
- More idiomatic usage of the entire project since it fits into the expected way of using dependencies

### Terraform Cloud Workflows

There's this idea floating around to support a workflow which is fully backed by the Terraform Cloud. This means, a user wouldn't even have to have the Terraform CLI installed to use `cdktf` (quite similar to the AWS CDK Cloudformation workflow)  That's only possible when we have prepackaged providers.

### Usage Information

Having prepackaged providers would also convey a bit of usage information via npm / python package download stats

## Conclusion 

Since prepackaged providers would just be built via `cdktf get`, the effort of creating these packages is quite low from my point of view. While it'll required being opinionated about versioning, I believe the provided value would outweigh potential drawbacks.

Would love to get some feedback on this!

**Comment by eladb** [comment]
My gut feeling is that this is going to blow up in complexity very quickly and I would recommend *not* publishing pre-built providers. I think it shouldn't be too hard to define a protocol which allows libraries to communicate which provider versions they need to operate and have the app pick that up. The app wouldn't need to generate the provider code itself, just to configure so it will be downloaded and available to the client during runtime.

**Comment by skorfmann** [comment]
> My gut feeling is that this is going to blow up in complexity very quickly and I would recommend _not_ publishing pre-built providers. 

Where do you see the complexity blowing up, could you elaborate on this a bit?

> I think it shouldn't be too hard to define a protocol which allows libraries to communicate which provider versions they need to operate and have the app pick that up. The app wouldn't need to generate the provider code itself, just to configure so it will be downloaded and available to the client during runtime.

You mean something along the lines of what we have with `cdktf.json`?

**Comment by jsteinich** [comment]
> The app wouldn't need to generate the provider code itself, just to configure so it will be downloaded and available to the client during runtime.

I'm not certain that will work on all potential languages. Seems like something that would fail with build time linking.

I agree that `cdktf get` is a bit of an awkward step for getting starting. So is having to use `cdktf.json` for typical use.

Some complexity I see comes down the road when/if there are multiple supported versions of cdktf. There could end being a matrix of packages needed. However, at some point there will probably need to be a line of what is officially supported and what is unofficial. Any provider that is officially supported will need to built with each version to ensure that nothing breaks anyways, so it doesn't seem like too much extra effort to also publish.

**Comment by skorfmann** [comment]
> Some complexity I see comes down the road when/if there are multiple supported versions of cdktf. There could end being a matrix of packages needed. 

I would keep this simple and strictly focus on `latest ` for both, the provider schema (e.g. `AWS`) and `cdktf`. Again, I think that's what most people would go for anyway.

**Comment by skorfmann** [comment]
One scenario which could become a bit complex:

When using multiple providers (e.g. `aws`, `null`, `google`) they have to share the exact same version of `cdktf`.  Essentially, `cdktf` would be a peer dependency and needs to be pinned to the respective major version of it.

But still, when there's a new major release of `ckdtf` all providers have to be bumped. Shouldn't be a huge problem.

**Comment by eladb** [comment]
> I'm not certain that will work on all potential languages. Seems like something that would fail with build time linking.

Why is this related to build? I might be missing something... All we need is for the correct provider version to be declared inside the terraform config and the resources defined in the library to reference it. A terraform config can have any number of providers defined in it. Even multiple versions of the same provider type.

**Comment by eladb** [comment]
> Where do you see the complexity blowing up, could you elaborate on this a bit?

Versioning, compatibility between apps and libraries, things like that.

> You mean something along the lines of what we have with cdktf.json?

No, I think something like construct metadata associated with the resource that is read by the CLI/app and used to synthesize the provider clauses.

**Comment by eladb** [comment]
Basically every resource should have a construct metadata associated with it which indicates which provider+version it requires and then during synthesis, the App can reflect on this metadata and create the right associations.

**Comment by jsteinich** [comment]
There are two main pieces of using a provider:

1. What gets included in the generated Terraform JSON
2. The actual code (TS/python/etc) for the resources/data sources/etc that the provider includes

The first part is easily solved with construct metadata / user provided values.

The second part is where things get trickier. Currently the user is required to put in a file what versions they want to use and then run a command to generate code. IMO that isn't the best experience, but setting that aside, bringing in "L2" concepts requires libraries that depend on providers. That isn't just a terraform dependency, there is a code dependency as well. Trying to delay that dependency resolution and/or making the user fill it in seems problematic.

**Comment by XBeg9** [comment]
I agree with @skorfmann, `cdktf` should eject `.gen` outside of individual projects, first of all you will require the community to commit those files, or this will become a bottleneck in dependency management across team members.

**Comment by XBeg9** [comment]
Due to this:
```
Users which require a specific provider version for some reason, will have to fallback to generate it manually by cdktf get
```

I tried to brainstorm around how we can keep Terraform provider (go package) to be consistent with the npm package, but let users use their own version of `cdktf` to generate provider.

- use @cdktf/provider-azure@exact_version
- put `cdktf` as peerDependency
- add `postinstall` npm hook, whenever this package installs, we will get .gen files. (we have the same for python) https://stackoverflow.com/questions/20288711/post-install-script-with-python-setuptools

I can make a sample project demonstrating this approach. What do you think? @skorfmann

**Comment by XBeg9** [comment]
You can test it here: https://github.com/XBeg9/cdktf-provider-docker, `@skitsko/cdktf-provider-docker@latest` and you will get compiled with your `cdktf-cli` version.

**Comment by XBeg9** [comment]
@skorfmann if you are fine with my approach, I can finish the full flow and create a PR.

**Comment by jsteinich** [comment]
@XBeg9 I don't think that approach will work well for other languages since not all package managers have that type of hook.

**Comment by XBeg9** [comment]
@jsteinich you are right, I've checked only `Python` and `Typescript`, which is currently officially supported. I don't know if there are any other plans to enable other languages.

**Comment by skorfmann** [comment]
@XBeg9 thanks for your suggestion, I'd love if we'd could do this. Unfortunately, I see the same issue as @jsteinich. The goal is supporting all languages which are provided by [jsii](https://github.com/aws/jsii) (Python, Java, C# would be possible today and likely Go at some point).

Besides from that, I don't think that `postInstall` for a fixed version would cut it in this case. You'd essentially have to do a peer dependency on `cdktf` in version `*` to never have to update the version dependency on that package. But even if that would be something we'd consider, how would the upgrade path look like? I think it would be bumping up the peer dependency in your project, removing `node_modules` and reinstall everything to get newly generated constructs. In summary the workflow would be still a bit clunky from my point view.

@XBeg9 out of curiosity, how many cases have you experienced where you actually had to pin a provider version in Terraform?

**Comment by XBeg9** [comment]
@skorfmann zero :) yea, I thought to not do fixed version, but anyway... I think we've agreed, that my approach could cause some side effects, and especially we don't want to lock us to specific `postInstall` hook, specifically to other languages.

I was not sure why you are taking each provider into isolated "monorepo", probably I need to understand the metadata and construct pieces. Can you expand a little bit here @skorfmann ?

Thanks!

**Comment by skorfmann** [comment]
> I was not sure why you are taking each provider into isolated "monorepo", probably I need to understand the metadata and construct pieces. Can you expand a little bit here @skorfmann ?

I think if you had a monorepo which contains all providers, it would be pretty hard to keep track of what's going on. For the custom constructs, I'm not sure if this would be really a thing, certainly not in the beginning. 

So, it'd probably be a dedicated repo per provider.

> zero :)

I had two occasions in a two year long project where we were heavily using Terraform. One was related to some sort of bug in the AWS provider and the other one a bug as well, but in the postgres provider. Both were fixed rather quickly, so we were able to track latests again.
